### PR TITLE
AWS Fargate gRPC cluster example

### PR DESCRIPTION
**Pull request by igorshmukler**

### CODE REVIEW

**Comment by igorshmukler** [comment]
I have read the CLA Document and I hereby sign the CLA

File: aws-fargate/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM authzed/spicedb:v1.12.0
+
+ARG SPICEDB_GRPC_PRESHARED_KEY
+ENV SPICEDB_GRPC_PRESHARED_KEY=${SPICEDB_GRPC_PRESHARED_KEY}
+ARG SPICEDB_DATASTORE_ENGINE
+ENV SPICEDB_DATASTORE_ENGINE=${SPICEDB_DATASTORE_ENGINE}
+ARG SPICEDB_DATASTORE_CONN_URI
+ENV SPICEDB_DATASTORE_CONN_URI=${SPICEDB_DATASTORE_CONN_URI}
```
Comment (vroldanbet) [inline aws-fargate/Dockerfile:8]: 🤔 it's not clear to me why this has to be added. Why can't those be specified as environment variables that get injected in the container via the cloudformation template? This container definition does not seem to provide anything the upstream couldn't provide

File: aws-fargate/fargate.yaml [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    Type: Number
+    Default: 1
+  # for autoscaling
+  MaxContainers:
+    Type: Number
+    Default: 2
+  # target CPU utilization (%)
+  AutoScalingTargetValue:
+    Type: Number
+    Default: 50
```
Comment (vroldanbet) [inline aws-fargate/fargate.yaml:49]: autoscaling could be disruptive to a spicedb cluster, because the ring has to reconfigure and it would potentially affect your deployment's cache hit ratios

File: aws-fargate/fargate.yaml [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+            - !Ref ContainerSecurityGroup
+      LoadBalancers:
+        - ContainerName: !Ref ServiceName
+          ContainerPort: !Ref ContainerPort
+          TargetGroupArn: !Ref TargetGroup
+  TargetGroup:
+    Type: AWS::ElasticLoadBalancingV2::TargetGroup
+    Properties:
+      HealthCheckIntervalSeconds: 300
+      # will look for a 200 status code by default unless specified otherwise
```
Comment (vroldanbet) [inline aws-fargate/fargate.yaml:195]: The configuration seems to be targeting the GRPC healthcheck, unless I'm missing something. The `ProtocolVersion` is specified as `GRPC` and the matcher is looking for a GrpcCode=0.

Would you mind clarifying this comment and those indicating gRPC specific conf is not tested?

**Review by vroldanbet — COMMENTED** [review]
Thanks for your contribution @igorshmukler! I left some questions 🙇🏻

File: aws-fargate/Dockerfile (hunk 1 above)
Comment (igorshmukler) [reply aws-fargate/Dockerfile:8]: sorry, i don't understand the comment. don't know the project as well as you probably do. is the goal of your question to understand whether my approach is the only approach possible? please clarify.

Comment (vroldanbet) [reply aws-fargate/Dockerfile:8]: Let me try again 😄 What I'm trying to say is that theoretically you wouldn't need to define your own `Dockerfile` and use the images we the Authzed team push to public registries like Dockerhub, Quay or GitHub Container Registry. My goal is to see if we can remove this `Dockerfile` and instead adjust the cloudformation accordingly to inject the corresponding environment variables. Does that make sense?

Comment (igorshmukler) [reply aws-fargate/Dockerfile:8]: Yes, you are certainly welcome to pull my PR and adjust the CloudFormation template. There are also other issues that are worth addressing, for example, parametrization of the ecr repository, certificate, aws account id, etc. I started the `TODO` section in the README. Please add anything you see fit there, as well as push further improvements.

This was meant as a PoC reference implementation. I would certainly be delighted to pull any upstream changes.

File: aws-fargate/fargate.yaml (hunk 2 above)
Comment (vroldanbet) [reply aws-fargate/fargate.yaml:49]: My initial suggestion was not to enable autoscaling. While SpiceDB supports adding and removing nodes, it's something that comes at a cost of reorganizing the hash ring, and leads to lower cache hit rate. However, after discussing with the team, there are things we could explore to reduce the impact of instances of SpiceDB coming and going, and we certainly want to make autoscaling fully supported without any performance impact.

So you can dismiss my comment!

File: aws-fargate/Dockerfile (hunk 1 above)
Comment (vroldanbet) [reply aws-fargate/Dockerfile:8]: Understood! I think the dockerfile bits is at least not something in line with our own standards, and it's only fair we tackle those. Another thing is dispatch, which isn't really enabled. Thanks for your contribution! As soon as we get some spare cycles we will get back to this 🙇🏻

Comment (igorshmukler) [reply aws-fargate/Dockerfile:8]: Yes, I do understand. You are correct. Appreciate the fact that this is a draft rather than a finished production-quality template, and it does require someone to co-pilot before the merge.

Perhaps in the future, I will get to the point where I will be able to cover an entire issue by myself, all the way over the finish line. Right now, I built MAYBE somewhat useful pieces. However, I don't have enough mileage to see every angle, like the core team members. Just sharing small bits that I developed while trying to get the initial SpiceDB integration functional enough to cover some of my use cases.
